std/os.zig

```diff
@@ -1,495 +1,477 @@
-
-pub const FileHandle = if (is_windows) windows.HANDLE else i32;
-
-pub const getAppDataDir = @import("os/get_app_data_dir.zig").getAppDataDir;
-pub const GetAppDataDirError = @import("os/get_app_data_dir.zig").GetAppDataDirError;
-
-const debug = std.debug;
-const assert = debug.assert;
-const testing = std.testing;
-
-const c = std.c;
-
-const mem = std.mem;
-const Allocator = mem.Allocator;
-
-const BufMap = std.BufMap;
-const cstr = std.cstr;
+pub use @import("os/bits.zig");
+
+/// See also `getenv`. Populated by startup code before main().
+pub var environ: [][*]u8 = undefined;
+
+/// Populated by startup code before main().
+/// Not available on Windows. See `std.process.args`
+/// for obtaining the process arguments.
+pub var argv: [][*]u8 = undefined;
+
+/// To obtain errno, call this function with the return value of the
+/// system function call. For some systems this will obtain the value directly
+/// from the return code; for others it will use a thread-local errno variable.
+/// Therefore, this function only returns a well-defined value when it is called
+/// directly after the system function call which one wants to learn the errno
+/// value of.
+pub const errno = system.getErrno;
+
+/// Closes the file descriptor.
+/// This function is not capable of returning any indication of failure. An
+/// application which wants to ensure writes have succeeded before closing
+/// must call `fsync` before `close`.
+/// Note: The Zig standard library does not support POSIX thread cancellation.
+pub fn close(fd: fd_t) void {
+    if (windows.is_the_target and !builtin.link_libc) {
+        return windows.CloseHandle(fd);
+    }
+    if (wasi.is_the_target) {
+        switch (wasi.fd_close(fd)) {
+            0 => return,
+            else => |err| return unexpectedErrno(err),
+        }
+    }
+    switch (errno(system.close(fd))) {
+        EBADF => unreachable, // Always a race condition.
+        EINTR => return, // This is still a success. See https://github.com/ziglang/zig/issues/2425
+        else => return,
+    }
+}
 
-const io = std.io;
-const base64 = std.base64;
-const ArrayList = std.ArrayList;
-const Buffer = std.Buffer;
-const math = std.math;
+pub const GetRandomError = OpenError;
 
-/// Fills `buf` with random bytes. If linking against libc, this calls the
+/// Obtain a series of random bytes. These bytes can be used to seed user-space
+/// random number generators or for cryptographic purposes.
+/// When linking against libc, this calls the
 /// appropriate OS-specific library call. Otherwise it uses the zig standard
 /// library implementation.
-pub fn getRandomBytes(buf: []u8) !void {
-    switch (builtin.os) {
-        Os.linux => while (true) {
-            // TODO check libc version and potentially call c.getrandom.
-            // See #397
-            const errno = posix.getErrno(posix.getrandom(buf.ptr, buf.len, 0));
-            switch (errno) {
+pub fn getrandom(buf: []u8) GetRandomError!void {
```

Comment:
I think this function should be in `std/crypto`?

Reply:
`std.crypto.randomBytes` calls (or is aliased to) this function. This exists in `std.os` because it's a Linux API and so this is the code that bridges libc and non-libc. Since it's a useful API to make cross platform, this is also the layer where it is made so, just like the other posix functions. You can think of it somewhat like `kevent` in that it's not POSIX but it is cross-platform.

Reply:
> Since it's a useful API to make cross platform, this is also the layer where it is made so, just like the other posix functions

I don't follow why it belongs `std/os` instead of  `std/crypto` though?

Reply:
> You can think of it somewhat like kevent in that it's not POSIX but it is cross-platform.

`kevent` is very not cross platform; there are many `kevent` semantics that are impossible to implement on linux,  windows or wasi.

Reply:
The OS provides some assurances about random numbers that a pure user-space implementation cannot (most importantly that the PRNG will be seeded).